libvips/foreign/webpsave.c

```diff
@@ -109,15 +155,639 @@ typedef struct _VipsForeignSaveWebp {
+G_DEFINE_ABSTRACT_TYPE( VipsForeignSaveWebP, vips_foreign_save_webp,
 	VIPS_TYPE_FOREIGN_SAVE );
 
+static void
+vips_webp_write_unset( VipsForeignSaveWebP *write )
+{
+	WebPMemoryWriterClear( &write->memory_writer );
+	VIPS_FREEF( WebPAnimEncoderDelete, write->enc );
+	VIPS_FREEF( WebPMuxDelete, write->mux );
+	VIPS_UNREF( write->image );
+}
+
+static void
+vips_foreign_save_webp_dispose( GObject *gobject )
+{
+	VipsForeignSaveWebP *webp= (VipsForeignSaveWebP *) gobject;
+
+	VIPS_UNREF( webp->frame );
+
+	VIPS_UNREF( webp->target );
+
+	VIPS_FREE( webp->frame_bytes );
+
+	G_OBJECT_CLASS( vips_foreign_save_webp_parent_class )->
+		dispose( gobject );
+}
+
+static gboolean
+vips_webp_pic_init( VipsForeignSaveWebP *write, WebPPicture *pic )
+{
+	if( !WebPPictureInit( pic ) ) {
+		vips_error( "webpsave", "%s", _( "picture version error" ) );
+		return( FALSE );
+	}
+	pic->writer = WebPMemoryWrite;
+	pic->custom_ptr = (void *) &write->memory_writer;
+
+	/* Smart subsampling needs use_argb because it is applied during
+	 * RGB to YUV conversion.
+	 */
+	pic->use_argb = write->lossless ||
+		write->near_lossless ||
+		write->smart_subsample;
+
+	return( TRUE );
+}
+
+/* Write a VipsImage into an unintialised pic.
+ */
+static int
+write_webp_image( VipsForeignSaveWebP *write, const VipsPel *imagedata, WebPPicture *pic )
+{
+	webp_import import;
+	int page_height = vips_image_get_page_height( write->image );
+
+	if( !vips_webp_pic_init( write, pic ) )
+		return( -1 );
+
+	pic->width = write->image->Xsize;
+	pic->height = page_height;
+
+	if( write->image->Bands == 4 )
+		import = WebPPictureImportRGBA;
+	else
+		import = WebPPictureImportRGB;
+
+	if( !import( pic, imagedata,
+		write->image->Xsize * write->image->Bands ) ) {
+		WebPPictureFree( pic );
+		vips_error( "webpsave", "%s", _( "picture memory error" ) );
+		return( -1 );
+	}
+
+	return( 0 );
+}
+
+/* We have a complete frame --- write!
+ */
+static int
+vips_foreign_save_webp_write_frame( VipsForeignSaveWebP *webp)
+{
+	WebPPicture pic;
+	int y;
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( webp );
+	VipsRect *frame_rect = &webp->frame->valid;
+	int page_index = frame_rect->top / frame_rect->height;
+
+	/* We need the frame as a contiguous RGB(A) buffer for libwebp.
+	 */
+	for( y = 0; y < frame_rect->height; y++ )
+		memcpy( webp->frame_bytes + y * webp->image->Bands * frame_rect->width,
+			VIPS_REGION_ADDR( webp->frame, 0, frame_rect->top + y ),
+			webp->image->Bands * frame_rect->width );
+
+	if( write_webp_image( webp, webp->frame_bytes, &pic ) ) {
+		return( -1 );
+	}
+
+	if( !WebPAnimEncoderAdd( webp->enc,
+		&pic, webp->timestamp_ms, &webp->config ) ) {
+		WebPPictureFree( &pic );
+		vips_error( class->nickname,
+			"%s", _( "anim add error" ) );
+		return( -1 );
+	}
+
+	/* Adjust current timestamp
+	 */
+	if( webp->delay &&
+		page_index < webp->delay_length )
+		webp->timestamp_ms += webp->delay[page_index] <= 10 ?
+			100 : webp->delay[page_index];
```

Comment:
Ooof that's a bit ugly. How about moving this to be next to the handling of small `gif_delay`? So have a small loop at line 635 squashing out low values.